src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2StreamWindowManager.cs

```diff
@@ -204,6 +209,14 @@ internal void OnDataOrHeadersReceived(Http2Connection connection)
                 {
                     if (initial) _initialBurst--;
 
+                    // When sendWindowUpdateBeforePing is true, try to send a WINDOW_UPDATE to make Google backends happy.
+                    // Unless we are doing the initial burst, do not send PING if we were not able to send the WINDOW_UPDATE.
+                    // See point 3. in the comments above the class definition for more info.
+                    if (sendWindowUpdateBeforePing && !connection.ForceSendConnectionWindowUpdate() && !initial)
```

Comment:
Where are places that can send a windows update? Can a connection be in the situation where a window update is always being sent somewhere else and when it comes time to send a ping, the pending window update is always empty?

Reply:
> Where are places that can send a windows update? 

For connection window the standard method is [`ExtendWindow`](https://github.com/dotnet/runtime/blob/7064b1fdc181ac56212f2de5eeefe8fe8809d40b/src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs#L1770C9-L1793) the new one introduced in this PR is [`ForceSendConnectionWindowUpdate`](https://github.com/dotnet/runtime/blob/7064b1fdc181ac56212f2de5eeefe8fe8809d40b/src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs#L1795C22-L1801). There are no other places sending WINDOW_UPDATE for the connection.

> Can a connection be in the situation where a window update is always being sent somewhere else, the pending window update is always empty?

`ExtendWindow` will send a "standard" connection WINDOW_UPDATE only if receiving DATA will make `_pendingWindowUpdate` pass `ConnectionWindowThreshold`. If that's the case, `sendWindowUpdateBeforePing` will be `false` at this line, so we won't do anything in `DataOrHeadersReceived`.

Note that receiving DATA triggers `ExtendWindow` and `DataOrHeadersReceived` sequentially:

https://github.com/dotnet/runtime/blob/7064b1fdc181ac56212f2de5eeefe8fe8809d40b/src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs#L772-L779

The PR changed the order of the two calls so the standard WINDOW_UPDATE goes before PING.